aquadoggo/src/libp2p/config.rs

```diff
@@ -0,0 +1,163 @@
+    /// exceeded, the ConnectionHandler will sleep.
+    pub per_connection_event_buffer_size: usize,
+
+    /// Ping behaviour enabled.
+    pub ping: bool,
+}
+
+impl Default for Libp2pConfiguration {
+    fn default() -> Self {
+        // Define the default listening multiaddress using the default QUIC port
+        let listening_multiaddr = format!("/ip4/0.0.0.0/udp/{QUIC_PORT}/quic-v1")
+            .parse()
+            .unwrap();
+
+        Self {
+            dial_concurrency_factor: 8,
+            listening_multiaddr,
+            max_connections_in: 16,
+            max_connections_out: 16,
+            max_connections_pending_in: 8,
+            max_connections_pending_out: 8,
+            max_connections_per_peer: 8,
+            mdns: false,
+            notify_handler_buffer_size: 128,
+            per_connection_event_buffer_size: 8,
+            ping: false,
+        }
+    }
+}
+
+impl Libp2pConfiguration {
+    /// Define the connection limits of the swarm.
+    pub fn connection_limits(&self) -> ConnectionLimits {
+        ConnectionLimits::default()
+            .with_max_pending_outgoing(Some(self.max_connections_pending_out))
+            .with_max_pending_incoming(Some(self.max_connections_pending_in))
+            .with_max_established_outgoing(Some(self.max_connections_out))
+            .with_max_established_incoming(Some(self.max_connections_in))
+            .with_max_established_per_peer(Some(self.max_connections_per_peer))
+    }
+
+    /// Load the keypair from the file at the specified path.
+    /// If the file does not exist, a random keypair is generated and saved.
+    /// If no path is specified, a random keypair is generated.
+    pub fn load_or_generate_keypair(path: Option<PathBuf>) -> Result<Keypair> {
+        let keypair = match path {
+            Some(mut path) => {
+                // Extend the base data directory path with the keypair filename
+                path.push(KEYPAIR_FILE_NAME);
+
+                // Check if the keypair file exists.
+                // If not, generate a new keypair and write it to file
+                if !path.is_file() {
+                    let identity: Keypair = Identity::new();
+                    identity.save(&path)?;
+                    info!("Created new libp2p keypair and saved it to {:?}", path);
+                    identity.keypair()
+                } else {
+                    // If the keypair file exists, open it and load the keypair
+                    let stored_identity: Keypair = Identity::load(&path)?;
+                    stored_identity.keypair()
+                }
+            }
+            None => {
+                // No path was provided. Generate and return a random keypair
+                Identity::new()
+            }
+        };
+
+        Ok(keypair)
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use super::Libp2pConfiguration;
+    use tempfile::TempDir;
```

Comment:
Mini style nitpick: We organise imports in that order and with a newline in-between: 1. external 2. crate 3. super (while we're using super mostly only in tests).